### PR TITLE
[mlir][SCF] Modernize `coalesceLoops` method to handle `scf.for` loops with iter_args

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
As part of this extension this change also does some general cleanup

1) Make all the methods take `RewriterBase` as arguments instead of
   creating their own builders that tend to crash when used within
   pattern rewrites
2) Split `coalesePerfectlyNestedLoops` into two separate methods, one
   for `scf.for` and other for `affine.for`. The templatization didnt
   seem to be buying much there.

Also general clean up of tests.

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 1]
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
...
+// in it modulo the number of iterations in this loop (remove the values
+// related to the outer loops):
+//   iv_i = floordiv(iv_linear, product-of-loop-ranges-until-i) mod range_i.
+// Compute these iteratively from the innermost loop by creating a "running
+// quotient" of division by the range.
+static std::pair<SmallVector<Value>, SmallPtrSet<Operation *, 2>>
+delinearizeInductionVariable(RewriterBase &rewriter, Location loc,
+                             Value linearizedIv, ArrayRef<Value> ubs) {
+  auto isIndexType = [](Value v) { return v.getType().isa<IndexType>(); };
+  if (linearizedIv.getType().isa<IndexType>()) {
```
Comment (qedawkins) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:619]:
```suggestion
  if (linearizedIv.getType().isIndex()) {
```

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 2]
```diff
@@ -557,57 +652,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
...
+    outerLoop.getBody()->getOperations().splice(
+        Block::iterator(innerLoop), innerLoop.getBody()->getOperations());
+    rewriter.replaceOp(innerLoop, yieldedVals);
+  }
   return success();
 }
 
+LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+  if (loops.empty()) {
+    return success();
```
Comment (qedawkins) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:709]: This looks inconsistent with the variant that take the rewriter, which returns failure if there are no loops
```
  if (loops.size() < 2)
    return failure();
```
This should either return failure too or the other should return success given no loops.

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir [hunk 3]
```diff
@@ -96,3 +101,150 @@ module attributes {transform.with_named_sequence} {
...
+    %0 = transform.structured.match ops{["scf.for"]} attributes {coalesce} in %arg1 : (!transform.any_op) -> !transform.any_op
+    %1 = transform.cast %0 : !transform.any_op to !transform.op<"scf.for">
+    %2 = transform.loop.coalesce %1 : (!transform.op<"scf.for">) -> (!transform.op<"scf.for">)
+    transform.yield
+  }
+}
+//  CHECK-DAG: #[[MAP:.+]] = affine_map<()[s0, s1, s2] -> ((-s0 + s1) ceildiv s2)>
+//  CHECK-DAG: #[[MAP1:.+]] = affine_map<()[s0, s1, s2, s3, s4, s5, s6, s7, s8] -> ((((-s0 + s1) ceildiv s2) * ((-s3 + s4) ceildiv s5)) * ((-s6 + s7) ceildiv s8))>
+//  CHECK-DAG: #[[MAP3:.+]] = affine_map<()[s0, s1, s2] -> (s0 * s1 + s2)>
+//      CHECK: func.func @tensor_loops(
```
Comment (qedawkins) [inline mlir/test/Dialect/SCF/transform-op-coalesce.mlir:132]: nit: CHECK-LABEL

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir [hunk 4]
```diff
@@ -96,3 +101,150 @@ module attributes {transform.with_named_sequence} {
...
+  return %0#0, %0#1 : tensor<?x?xf32>, tensor<?x?xf32>
+}
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["scf.for"]} attributes {coalesce} in %arg1 : (!transform.any_op) -> !transform.any_op
+    %1 = transform.cast %0 : !transform.any_op to !transform.op<"scf.for">
+    %2 = transform.loop.coalesce %1 : (!transform.op<"scf.for">) -> (!transform.op<"scf.for">)
+    transform.yield
+  }
+}
```
Comment (qedawkins) [inline mlir/test/Dialect/SCF/transform-op-coalesce.mlir:182]: ditto: Can you add a CHECK-LABEL here and below? It helps debugging confusing CHECK-NOT related lit test failures (and you can drop the checks for `transform.named_sequence` as well, because I assume that's why those checks are there).

**Review by qedawkins — APPROVED** [review]
LGTM after addressing comments. Might be good to give someone else a chance to check that the new Rewriter API function makes sense, I don't have any context there.

File: mlir/lib/Dialect/Affine/Transforms/LoopCoalescing.cpp [hunk 5]
```diff
@@ -35,13 +35,17 @@ namespace {
 struct LoopCoalescingPass
     : public affine::impl::LoopCoalescingBase<LoopCoalescingPass> {
 
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect>();
+  }
```
Comment (ftynse) [inline mlir/lib/Dialect/Affine/Transforms/LoopCoalescing.cpp:40]: Why is this necessary? Can this pass produce entities from the affine dialect when none were present in the input?

File: mlir/lib/Dialect/Affine/Utils/LoopUtils.cpp [hunk 6]
```diff
@@ -2765,3 +2765,49 @@ mlir::affine::separateFullTiles(MutableArrayRef<AffineForOp> inputNest,
...
+LogicalResult affine::coalescePerfectlyNestedAffineLoops(AffineForOp op) {
+  LogicalResult result(failure());
+  SmallVector<AffineForOp> loops;
+  getPerfectlyNestedLoops(loops, op);
+
+  // Look for a band of loops that can be coalesced, i.e. perfectly nested
+  // loops with bounds defined above some loop.
+  // 1. For each loop, find above which parent loop its operands are
+  // defined.
+  SmallVector<unsigned, 4> operandsDefinedAbove(loops.size());
```
Comment (ftynse) [inline mlir/lib/Dialect/Affine/Utils/LoopUtils.cpp:2778]: Nit: drop the explicit number of stack elements.

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp [hunk 7]
```diff
@@ -8,6 +8,7 @@
 
 #include "mlir/Dialect/SCF/Transforms/Passes.h"
 
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp:11]: Why is this necessary?

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp [hunk 8]
```diff
@@ -28,6 +29,11 @@ namespace {
 struct TestSCFParallelLoopCollapsing
     : public impl::TestSCFParallelLoopCollapsingBase<
           TestSCFParallelLoopCollapsing> {
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect>();
+  }
+
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp:36]: We shouldn't have a dependency on affine from SCF.

File: mlir/include/mlir/Dialect/SCF/Utils/Utils.h [hunk 9]
```diff
@@ -100,11 +100,16 @@ getSCFMinMaxExpr(Value value, SmallVectorImpl<Value> &dims,
 /// `loops` contains a list of perfectly nested loops with bounds and steps
 /// independent of any loop induction variable involved in the nest.
 LogicalResult coalesceLoops(MutableArrayRef<scf::ForOp> loops);
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/SCF/Utils/Utils.h:102]: Should we remove the version that doesn't take the rewriter?

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 10]
```diff
@@ -12,7 +12,9 @@
 
 #include "mlir/Dialect/SCF/Utils/Utils.h"
 #include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:15]: We shouldn't depend on affine from scf.

**Review by ftynse — CHANGES_REQUESTED** [review]
SCF must not depend on Affine.

**Comment by MaheshRavishankar** [comment]
> SCF must not depend on Affine.

I see your point... but I was trying to use `affine.apply` and `affine.delinearize_index`. Thats why the Affine dialect registration and the dependence come in. Should we just drop that? That would be kind of unfortunate.

File: mlir/include/mlir/Dialect/SCF/Utils/Utils.h (hunk 9 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/SCF/Utils/Utils.h:102]: I am fine doing that. Should be a simple fix for downstream users.

Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/SCF/Utils/Utils.h:102]: Actually tried it, but this has some strange coupling with things in Affine dialect. I am just going to leave it there for now for people to use at their own risk.

File: mlir/lib/Dialect/Affine/Transforms/LoopCoalescing.cpp (hunk 5 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/Transforms/LoopCoalescing.cpp:40]: Removed now.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 11]
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
   return success();
 }
 
+namespace {
+// When outer dimension used for delinearization are ones, the corresponding
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4536]: I think so, but wasnt hundred percent sure of the math, but you are right. Ill split this out of this PR though cause I dont need this right now.

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp (hunk 7 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp:11]: Removed.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 12]
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
...
-      isZeroBased
-          ? scaled
-          : insideLoopBuilder.create<arith::AddIOp>(loc, scaled, lowerBound);
-
-  SmallPtrSet<Operation *, 2> preserve{scaled.getDefiningOp(),
-                                       shifted.getDefiningOp()};
-  inductionVar.replaceAllUsesExcept(shifted, preserve);
-  return {/*lowerBound=*/newLowerBound, /*upperBound=*/newUpperBound,
-          /*step=*/newStep};
+      isStepOne ? diff : rewriter.create<arith::CeilDivSIOp>(loc, diff, step);
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:527]: It should. But doesnt matter now.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 13]
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
...
+                                  preserve.value());
+  } else {
+    rewriter.replaceAllUsesWith(normalizedIv, unNormalizedIv);
+  }
+  return;
 }
 
-LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+/// Helper function to multiply a sequence of values.
+static OpFoldResult getProductOfIndexes(RewriterBase &rewriter, Location loc,
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:580]: Removed uses of affine dialect ops here, so this isnt a problem anymore.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:709]: Good catch. Made both failure fo rnow.

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir (hunk 3 above)
Comment (MaheshRavishankar) [reply mlir/test/Dialect/SCF/transform-op-coalesce.mlir:132]: I leave it as CHECK for now. CHECK-LABEL resets the context that can be pierced in wierd ways. Also I think `CHECK-SAME` works with CHECK-LABEL, but I am not sure.

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir (hunk 4 above)
Comment (MaheshRavishankar) [reply mlir/test/Dialect/SCF/transform-op-coalesce.mlir:182]: Actually thats why added `CHECK-NOT` so that there is no match till the `transform.named_sequence`. Ill leave is as `CHECK` for now.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 14]
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineDelinearizeIndexOp indexOp,
+                                PatternRewriter &rewriter) const override {
+    ValueRange basis = indexOp.getBasis();
+    if (basis.empty()) {
+      return failure();
+    }
+    std::optional<int64_t> basisValue =
+        getConstantIntValue(getAsOpFoldResult(basis.front()));
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp]: I removed this from this commit. Ill resubmit that later and incorporate this suggestion.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 15]
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+    if (!basisValue || basisValue != 1) {
+      return failure();
+    }
+    SmallVector<Value> replacements;
+    Location loc = indexOp.getLoc();
+    replacements.push_back(rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    auto newIndexOp = rewriter.create<AffineDelinearizeIndexOp>(
+        loc, indexOp.getLinearIndex(), basis.drop_front());
+    replacements.append(newIndexOp->result_begin(), newIndexOp->result_end());
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp]: I removed this from this commit. Ill resubmit that later and incorporate this suggestion.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 16]
```diff
@@ -557,57 +591,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
...
+    auto innerLoop = loops[i];
+    if (outerloop.getNumRegionIterArgs() != innerLoop.getNumRegionIterArgs()) {
+      continue;
+    }
+    if (llvm::any_of(
+            llvm::zip_equal(outerloop.getRegionIterArgs(),
+                            innerLoop.getInitArgs()),
+            [](auto it) { return std::get<0>(it) != std::get<1>(it); })) {
+      continue;
+    }
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:697]: Nit: don't we have an `operator==` on `ValueRange` that we could use here?

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir [hunk 17]
```diff
@@ -96,3 +102,200 @@ module attributes {transform.with_named_sequence} {
...
+// CHECK-SAME:     %[[LB1:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[UB1:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[STEP1:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[LB2:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[UB2:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[STEP2:[a-zA-Z0-9_]+]]: index
+//      CHECK:   %[[NEWUB0_DIFF:.+]] = arith.subi %[[UB0]], %[[LB0]]
+//  CHECK-DAG:   %[[NEWUB0:.+]] = arith.ceildivsi %[[NEWUB0_DIFF]], %[[STEP0]]
+//  CHECK-DAG:   %[[C0:.+]] = arith.constant 0
+//  CHECK-DAG:   %[[C1:.+]] = arith.constant 1
```
Comment (ftynse) [inline mlir/test/Dialect/SCF/transform-op-coalesce.mlir:144]: Why is this a DAG? Normally, there shouldn't be any non-determinism in the output IR.

**Review by ftynse — APPROVED** [review]
Thanks for iterating! LGTM modulo some nits.

File: mlir/test/Dialect/SCF/transform-op-coalesce.mlir (hunk 17 above)
Comment (joker-eph) [reply mlir/test/Dialect/SCF/transform-op-coalesce.mlir:144]: DAG isn't about non-determinism (otherwise it wouldn't exist at all, since as you mention there shouldn't be any). Instead it can indicate portion where "we don't care about the order". In particular for constant, we have been changing the strategy of "CSE on the fly" quite a few times, and it's annoying when test encode an order while it isn't relevant to what is actually tested.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 18]
```diff
@@ -472,18 +473,23 @@ LogicalResult mlir::loopUnrollByFactor(
...
+/// Transform a loop with a strictly positive step
+///   for %i = %lb to %ub step %s
+/// into a 0-based loop with step 1
+///   for %ii = 0 to ceildiv(%ub - %lb, %s) step 1 {
+///     %i = %ii * %s + %lb
+/// Insert the induction variable remapping in the body of `inner`, which is
+/// expected to be either `loop` or another loop perfectly nested under `loop`.
+/// Insert the definition of new bounds immediate before `outer`, which is
+/// expected to be either `loop` or its parent in the loop nest.
+static LoopParams normalizeLoop(RewriterBase &rewriter, Location loc, Value lb,
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:485]: Actually it is doing it in place. So maybe normalizeLoops is better.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 19]
```diff
@@ -557,57 +591,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
...
+    for (auto [outerLoopIterArg, innerLoopIterArg] : llvm::zip_equal(
+             outerLoop.getRegionIterArgs(), innerLoop.getRegionIterArgs())) {
+      rewriter.replaceAllUsesExcept(innerLoopIterArg, outerLoopIterArg,
+                                    preservedUsers);
+    }
+    Operation *innerTerminator = innerLoop.getBody()->getTerminator();
+    auto yieldedVals = llvm::to_vector(innerTerminator->getOperands());
+    rewriter.eraseOp(innerTerminator);
+    outerLoop.getBody()->getOperations().splice(
+        Block::iterator(innerLoop), innerLoop.getBody()->getOperations());
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:640]: Good catch. Thanks!

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ftynse) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:485]: Hmm, I don't see _this_ function doing it. The caller (`coalesceLoops`) does the in-place around line 600.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp (hunk 16 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:697]: Yeah, its unfortunate cause these are BlockList, ValueRange, ArgList and OpResultList... so no direct equality here.

Comment (kuhar) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:697]: we have: `llvm::equal`, it should work here

Comment (ftynse) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:697]: I also suspect that all these are convertible to `ValueRange`, but we may need a explicit conversion constructor or something.
